application/common/models/User.php

```diff
@@ -189,17 +194,45 @@ public function rules(): array
     private function validatePassword(): Closure
     {
         return function ($attributeName) {
-            
+
             if ( ! $this->hasPasswordAlready()) {
                 $this->attemptPasswordMigration();
             }
-            
+
             if (! password_verify($this->password, $this->password_hash)) {
                 $this->addError($attributeName, 'Incorrect password.');
             }
         };
     }
 
+    private function validateExpiration(): Closure
+    {
+        return function ($attributeName) {
+            $now = time();
+            $expiration = $this->addGracePeriod($this->getPasswordExpiration());
+
+            if ($now > $expiration) {
+                $this->addError($attributeName, 'Expired password.');
+            }
+        };
+    }
+
+    private function getPasswordExpiration(): string
+    {
+        /** @var $mostRecentPassword PasswordHistory */
+        $mostRecentPassword = $this->getPasswordHistories()
+                                   ->orderBy(['id' => SORT_DESC])
+                                   ->one();
+
+        return $mostRecentPassword->expires();
+
+    }
+
+    private function addGracePeriod(string $expiration): int
```

Comment:
the expiration will come out with the `User` as we discussed.  That will be enough, correct @Baggerone ?